### ISSUE TITLE
VerifiedCache should be re-enabled

### ISSUE THREAD

**Issue by snape479** [mimblewimble/grin#3620]
Due to the recent CVE, the VerifierCache was disabled as an emergency measure. It should be enabled with the appropriate fix. PR will be forthcoming with this fix.

**Comment by antiochp** [comment]
Duplicate of https://github.com/mimblewimble/grin/issues/3606

**Comment by phyro** [comment]
I suggest we close this one

**Comment by snape479** [comment]
Added a perf test to https://github.com/mimblewimble/grin/pull/3621. It shows a significant performance impact from disabling the verifier_cache. Here is the output:

# No cache
% ./target/release/perf
[2021-04-02 08:02:54]: Starting perf test
[2021-04-02 08:02:55]: building tx1
[2021-04-02 08:03:42]: building tx2
[2021-04-02 08:04:28]: building tx3
[2021-04-02 08:05:14]: building tx4
[2021-04-02 08:06:01]: building tx5
[2021-04-02 08:06:47]: start validate tx
[2021-04-02 08:06:51]: tx val time elapsed = 3599ms
[2021-04-02 08:06:51]: start process block
[2021-04-02 08:06:52]: block val time elapsed = 909ms

# With cache
% ./target/release/perf
[2021-04-02 07:58:08]: Starting perf test
[2021-04-02 07:58:09]: building tx1
[2021-04-02 07:58:55]: building tx2
[2021-04-02 07:59:42]: building tx3
[2021-04-02 08:00:28]: building tx4
[2021-04-02 08:01:15]: building tx5
[2021-04-02 08:02:01]: start validate tx
[2021-04-02 08:02:02]: tx val time elapsed = 878ms
[2021-04-02 08:02:02]: start process block
[2021-04-02 08:02:02]: block val time elapsed = 231ms

I think you should reconsider closing out this issue and merge pull request #3621.

**Comment by antiochp** [comment]
@snape479 Thanks for digging into this. Let me take a look at the test results today.
I'm surprised we are seeing this large a difference.

**Comment by snape479** [comment]
Yes, the test uses a pathological case for the tx validation, as only one of the outputs is changed. Keep in mind an attacker may use the fact that they can make transactions that look almost the same that are actually different and use the pathological case because nodes would propagate these 5 transactions separately through the transaction pools whereas if they were exactly the same, they would only propagate as a single transaction. In terms of the block though, I think all the transactions will generally be cached and thus I think that test shows the actual picture.

**Comment by phyro** [comment]
Is the general consensus here that we add some form of caching later on once the benefits become relevant for our on-chain/mempool volume? if yes, I'd suggest we also consider updating/closing the linked PR https://github.com/mimblewimble/grin/pull/3621
